lib/balrog/middleware.rb

```diff
@@ -7,35 +7,58 @@
 # This is typically set up in an initialize when you run
 # `rails g balrog:install`, and looks a bit like this:
 #
-#    Rails.application.config.middleware.use Balrog::Middleware do
-#      password_hash '<bcrypt hash>'
-#    end
+#  Balrog::Middleware.setup do |config|
+#    config.set_password_hash '<bcrypt hash>'
+#  end
+
 class Balrog::Middleware
-  def initialize(app, &block)
+
+  mattr_reader :password_hash
+  mattr_reader :session_length
+  mattr_reader :omniauth_config
+  mattr_reader :domain_whitelist
+
+  def initialize(app)
     @app = app
-    instance_eval(&block) if block_given?
   end
 
   def call(env)
     path = env["PATH_INFO"]
     method = env["REQUEST_METHOD"]
-    if method == 'POST' && path == '/balrog/signin'
+    if login_request?(path, method)
       handle_login(env)
-    elsif method == "DELETE" && path == '/balrog/logout'
+    elsif omniauth_request?(path, method)
+      handle_omniauthentication(env)
+    elsif logout_request?(path, method)
       handle_logout(env)
     else
       @app.call(env)
     end
   end
 
+  def self.setup
+    yield self
+  end
+
   private
 
-  def password_hash(input)
-    @password_hash = BCrypt::Password.new(input)
+  def self.set_password_hash(input)
+    @@password_hash = BCrypt::Password.new(input)
+  end
+
+  def self.set_omniauth(provider, *args)
+    @@omniauth_config = {
+      provider: provider,
+      args: args
+    }
+  end
+
+  def self.set_domain_whitelist(*domains)
+    @@domain_whitelist = domains
   end
 
-  def set_session_expiry(time_period)
-    @session_length = time_period
+  def self.set_session_expiry(time_period)
+    @@session_length = time_period
   end
 
   def handle_login(env)
```

Comment:
All of these handlers should have some comments now, as it's starting to get more confusing knowing what each one does.